pkg/server/endpoints/harvester.go

```diff
@@ -0,0 +1,189 @@
+	"errors"
+	"fmt"
+	"io"
+
+	"github.com/HewlettPackard/galadriel/pkg/common"
+	"github.com/HewlettPackard/galadriel/pkg/common/util"
+	"github.com/labstack/echo/v4"
+	"github.com/spiffe/go-spiffe/v2/spiffeid"
+)
+
+func (e *Endpoints) syncFederatedBundleHandler(ctx echo.Context) error {
+	e.Log.Debug("Receiving sync request")
+
+	token, ok := ctx.Get("token").(*common.AccessToken)
+	if !ok {
+		err := errors.New("error asserting user token")
+		e.handleTcpError(ctx, err.Error())
+		return err
+	}
+	harvesterTrustDomain := token.TrustDomain
+
+	body, err := io.ReadAll(ctx.Request().Body)
+	if err != nil {
+		e.handleTcpError(ctx, fmt.Sprintf("failed to read body: %v", err))
+		return err
+	}
+
+	receivedHarvesterState := common.SyncBundleRequest{}
+	err = json.Unmarshal(body, &receivedHarvesterState)
+	if err != nil {
+		e.handleTcpError(ctx, fmt.Sprintf("failed to unmarshal state: %v", err))
+		return err
+	}
+
+	harvesterBundleDigests := receivedHarvesterState.State
+
+	_, foundSelf := receivedHarvesterState.State[token.TrustDomain]
+	if foundSelf {
+		e.handleTcpError(ctx, "bad request: harvester cannot federate with itself")
+		return err
+	}
+
+	relationships, err := e.DataStore.GetRelationships(context.TODO(), harvesterTrustDomain.String())
+	if err != nil {
+		e.handleTcpError(ctx, fmt.Sprintf("failed to fetch relationships: %v", err))
+		return err
+	}
+
+	federatedMembers := getFederatedMembers(relationships, harvesterTrustDomain)
+	lastBundlesDigests := getCurrentFederatedBundleDigests(federatedMembers)
+	bundlesUpdates := getFederatedBundlesUpdates(harvesterBundleDigests, federatedMembers)
+
+	response := common.SyncBundleResponse{}
+	response.State = lastBundlesDigests
+	response.Updates = bundlesUpdates
+
+	responseBytes, err := json.Marshal(response)
+	if err != nil {
+		e.handleTcpError(ctx, fmt.Sprintf("failed to marshal response: %v", err))
+		return err
+	}
+
+	_, err = ctx.Response().Write(responseBytes)
+	if err != nil {
+		e.handleTcpError(ctx, fmt.Sprintf("failed to write response: %v", err))
+		return err
+	}
+
+	return nil
+}
+
+func getFederatedMembers(relationships []*common.Relationship, td spiffeid.TrustDomain) []*common.Member {
+	federatedMembers := make([]*common.Member, 0)
+
+	for _, r := range relationships {
+		ma := r.MemberA
+		mb := r.MemberB
+
+		if td.Compare(ma.TrustDomain) != 0 && ma != nil {
+			federatedMembers = append(federatedMembers, ma)
+		} else if mb != nil {
+			federatedMembers = append(federatedMembers, mb)
+		}
+	}
+	return federatedMembers
+}
+
+func getFederatedBundlesUpdates(harvesterBundlesDigests common.BundlesDigests, federatedMembers []*common.Member) common.BundleUpdates {
+
+	response := make(common.BundleUpdates)
+
+	for _, m := range federatedMembers {
+		td := m.TrustDomain
+		serverDigest := m.BundleDigest
+		harvesterDigest := harvesterBundlesDigests[td]
+
+		// if the bundle digest the harvester has for a federated trust domain is not the same as the digest the server has,
```

Comment:
super nit: I think its more clear like this, took me a couple of minutes to understand the phrase xD. Feel free to ignore it.
⬇️ 
```suggestion
		// If the harvester has a bundle digest for a federated trust domain different from the the one the Server has,
```

Reply:
I've tried improving the whole thing, PTAL.